### ISSUE TITLE
Encrypt/decrypt attributes with casts or model property

### ISSUE THREAD

**Issue by barryvdh** [laravel/ideas#1244]
To make it easier to encrypt/decrypt columns in the database, it could be built-in to the Eloquent model, similar to either the `$dates` attribute or the `$casts` properties.
Using an extra property would allow for being cast to a type after decrypting, so perhaps better.

To make it easy for users to encrypt existing data, a command could be added to encrypt data based on these columns. (eg. `php artisan db:encrypt --class="App\User"`)

It could be done using mutators/accessors, but having this built-in could make it easier and cache objects to avoid multiple times decrypting the same value in the same request.

**Comment by 36864** [comment]
I like it.

We'd also need a way to re-encrypt old data with a new key/algorithm, something like `php artisan db::recrypt --class="App\User" --old-key=OLD_KEY --old-algorithm=bcrypt --new-key=NEW_KEY --new-algorithm=argon2`

Or perhaps a new configuration file with different encryption settings, such as:
```
//encryption.php
return [
  'users' => [
    'algorithm' =>  'bcrypt',
    'key' => env('USERS_BCRYPT_KEY'),
  ],
  'transactions => 
    'argon' => [
      'algorithm' => 'argon2',
      'key' => env('TRANSACTIONS_ARGON2_KEY'),
    ],
    'bcrypt' => [
      'algorithm' => 'bcrypt',
      'key' => env('TRANSACTIONS_ARGON2_KEY'),
    ],
  ],
];
```

which would then allow us to reference these settings:
```
php artisan db:encrypt --class="App\User" --config="encryption.users"

php artisan db:recrypt --class="App\Transactions" --from="encryption.transactions.bcrypt" --to="encryption.transactions.argon"
```

Although now that I've typed that all out, it would probably require a major rewrite of the encryption system.

**Comment by barryvdh** [comment]
Yeah that would be a bit more complicated. (re-)Encrypting the database could be as simple as getting the attributes and doing:

```php
foreach ($attributes as $attribute) {
	try {
		// Try to decrypt it using the default encryption
		$item->{$attribute} = $user->{$attribute};
	} catch (DecryptException $e) {
		// When fails, assume it's not encrypted yet and use the raw version
		$item->{$attribute} = $item->getOriginal($attribute);
	}
}
$item->save();
```

**Comment by 36864** [comment]
How would your code get the original value if it fails to decrypt it?

**Comment by barryvdh** [comment]
No that's what I said, that would be more complicated. I just meant that you could run the command again with additional properties encrypted, and it would encrypt the new values without breaking the existing ones.

**Comment by sisve** [comment]
I'm a bit worried that announcing "support for automatically encrypt and decrypt attributes" means someone will file a bug report since they cannot search the database for these fields.`User::where('encrypted_field', '=', $secretValue)->first()`. Is this out-of-scope for this issue?

**Comment by barryvdh** [comment]
Yes that is out of scope.

**Comment by poisa** [comment]
I'm willing to create a proof-of-concept PR if there is interest in this.

**Comment by andrey-helldar** [comment]
@barryvdh, it seems to me that the changes proposed in laravel/framework#30958 PR will allow this to be implemented in custom castes.
